Easymock does not work with Java 9 b148+

It is no longer possible to inject classes into foreign ClassLoaders by making `ClassLoader#defineClass()` accessible. This breaks CGLIB, that is used by EasyMock. The corresponding issue is https://github.com/cglib/cglib/issues/93. To inject classes into foreign ClassLoaders you have to use `sun.misc.Unsafe#defineClass()`, because Java 9 has a special case for legacy code, so you can still get a "deprecated backwards-compatibility Unsafe" (Java 9 internally uses another Unsafe).

See issue: https://issues.apache.org/jira/browse/SOLR-9893

Example stack trace (using Easymock 3.4 + Java 9 EA build 152):

```
   [junit4]    > Throwable #1: java.lang.ExceptionInInitializerError
   [junit4]    >        at __randomizedtesting.SeedInfo.seed([AC6240241F854C7:5B230C2B8AB54ED7]:0)
   [junit4]    >        at org.easymock.cglib.core.KeyFactory$Generator.generateClass(KeyFactory.java:166)
   [junit4]    >        at org.easymock.cglib.core.DefaultGeneratorStrategy.generate(DefaultGeneratorStrategy.java:25)
   [junit4]    >        at org.easymock.cglib.core.AbstractClassGenerator.create(AbstractClassGenerator.java:216)
   [junit4]    >        at org.easymock.cglib.core.KeyFactory$Generator.create(KeyFactory.java:144)
   [junit4]    >        at org.easymock.cglib.core.KeyFactory.create(KeyFactory.java:116)
   [junit4]    >        at org.easymock.cglib.core.KeyFactory.create(KeyFactory.java:108)
   [junit4]    >        at org.easymock.cglib.core.KeyFactory.create(KeyFactory.java:104)
   [junit4]    >        at org.easymock.cglib.proxy.Enhancer.<clinit>(Enhancer.java:69)
   [junit4]    >        at org.easymock.internal.ClassProxyFactory.createEnhancer(ClassProxyFactory.java:257)
   [junit4]    >        at org.easymock.internal.ClassProxyFactory.createProxy(ClassProxyFactory.java:165)
   [junit4]    >        at org.easymock.internal.MocksControl.createMock(MocksControl.java:113)
   [junit4]    >        at org.easymock.internal.MocksControl.createMock(MocksControl.java:94)
   [junit4]    >        at org.easymock.EasyMock.mock(EasyMock.java:69)
   [junit4]    >        at org.easymock.EasyMock.createMock(EasyMock.java:311)
   [junit4]    >        at org.apache.solr.core.CoreSorterTest.getMockContainer(CoreSorterTest.java:107)
   [junit4]    >        at org.apache.solr.core.CoreSorterTest.testSort(CoreSorterTest.java:87)
   [junit4]    >        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   [junit4]    >        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   [junit4]    >        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   [junit4]    >        at java.base/java.lang.reflect.Method.invoke(Method.java:543)
   [junit4]    >        at java.base/java.lang.Thread.run(Thread.java:844)
   [junit4]    > Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @3cfab390
   [junit4]    >        at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:207)
   [junit4]    >        at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:192)
   [junit4]    >        at java.base/java.lang.reflect.Method.setAccessible(Method.java:186)
   [junit4]    >        at org.easymock.cglib.core.ReflectUtils$2.run(ReflectUtils.java:56)
   [junit4]    >        at java.base/java.security.AccessController.doPrivileged(Native Method)
   [junit4]    >        at org.easymock.cglib.core.ReflectUtils.<clinit>(ReflectUtils.java:46)
   [junit4]    >        ... 54 more
```

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -35,20 +35,20 @@
     <dependency>
       <groupId>cglib</groupId>
       <artifactId>cglib</artifactId>
-      <version>3.2.4</version>
+      <version>3.2.5</version>
     </dependency>
-    <!-- Use version 5 to be compliant with Java 8 -->
+    <!-- Use version 6 to be compliant with Java 9 -->
     <dependency>
       <groupId>org.ow2.asm</groupId>
       <artifactId>asm</artifactId>
-      <version>5.1</version>
+      <version>6.0_BETA</version>
       <scope>runtime</scope>
     </dependency>
     <!-- Used for class mocking -->
     <dependency>
       <groupId>org.objenesis</groupId>
       <artifactId>objenesis</artifactId>
-      <version>2.4</version>
+      <version>2.6</version>
     </dependency>
     <!-- Used for class mocking on Android (cglib replacement) -->
     <dependency>

--- a/pom.xml
+++ b/pom.xml
@@ -131,7 +131,7 @@
         </plugin>
         <plugin>
           <artifactId>maven-assembly-plugin</artifactId>
-          <version>3.0.0</version>
+          <version>3.1.0</version>
           <executions>
             <execution>
               <id>easymock-bundle</id>
@@ -162,7 +162,7 @@
         </plugin>
         <plugin>
           <artifactId>maven-compiler-plugin</artifactId>
-          <version>3.5.1</version>
+          <version>3.6.2</version>
         </plugin>
         <plugin>
           <artifactId>maven-deploy-plugin</artifactId>
@@ -227,7 +227,7 @@
         <plugin>
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-shade-plugin</artifactId>
-          <version>2.4.3</version>
+          <version>3.1.0</version>
         </plugin>
         <!--This plugin configuration is used to store Eclipse m2e settings only. It has no influence on the Maven build itself. -->
         <plugin>

--- a/test-integration/pom.xml
+++ b/test-integration/pom.xml
@@ -30,7 +30,7 @@
     <jacoco.skip>true</jacoco.skip>
 
     <!-- Versions -->
-    <powermock.version>1.6.6</powermock.version>
+    <powermock.version>1.7.1</powermock.version>
   </properties>
 
   <dependencies>

